### PR TITLE
Users/sachinma/bugfix

### PR DESCRIPTION
**Pull request by sachinma**
Adding retry for ETIMEDOUT scenario

### CODE REVIEW

File: Tasks/Common/azure-arm-rest/webClient.ts [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+            }
+            else {
+                throw error;
+            }
+
+        }
+    }
+}
+
+function beginRequestInternal(request: WebRequest): Promise<WebResponse> {
```
Comment (ds-ms) [inline Tasks/Common/azure-arm-rest/webClient.ts]: Now that it's a part of webClient, we could name it as beginRequest maybe?

File: Tasks/Common/azure-arm-rest/AzureServiceClient.ts [hunk 2]
```diff
@@ -115,7 +103,7 @@ export class ServiceClient {
         return headers;
```
Comment (ds-ms) [inline Tasks/Common/azure-arm-rest/AzureServiceClient.ts]: hasOwnProperty is a redundant check here.

File: Tasks/Common/azure-arm-rest/webClient.ts [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+    public retriableErrorCodes: string[];
+    public retryCount: number;
+    public retryIntervalInSeconds: number;
+}
+
+export async function sendRequest(request: WebRequest, options?: WebRequestOptions): Promise<WebResponse> {
+    let i = 0;
+    let retryCount = options ? options.retryCount : 5;
+    let retryIntervalInSeconds = options ? options.retryIntervalInSeconds : 5;
+    let retriableErrorCodes = options ? options.retriableErrorCodes : ["ETIMEDOUT"];
```
Comment (ds-ms) [inline Tasks/Common/azure-arm-rest/webClient.ts]: Should we not mandate the retry on ETIMEDOUT? i.e. if retriableErrorCodes doesn't have ETIMEDOUT, you add it to the array

File: Tasks/Common/azure-arm-rest/AzureServiceClient.ts (hunk 2 above)
Comment (sachinma) [inline Tasks/Common/azure-arm-rest/AzureServiceClient.ts]: I have left it for now

File: Tasks/Common/azure-arm-rest/webClient.ts (hunk 3 above)
Comment (sachinma) [inline Tasks/Common/azure-arm-rest/webClient.ts]: If user is over-riding it - lets not add a extra value

File: Tasks/Common/azure-arm-rest/webClient.ts [hunk 4]
```diff
@@ -0,0 +1,86 @@
+import tl = require('vsts-task-lib/task');
+import util = require("util")
+var httpClient = require('vso-node-api/HttpClient');
```
Comment (bishal-pdMSFT) [inline Tasks/Common/azure-arm-rest/webClient.ts]: Not directly related to this PR, but why do we need vso-node-api dependency here? All we need is some http client. we can directly use what vso-node-api is using - i think it is `'typed-rest-client/HttpClient'`

File: Tasks/Common/azure-arm-rest/webClient.ts [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+    public retriableErrorCodes: string[];
+    public retryCount: number;
+    public retryIntervalInSeconds: number;
+}
+
+export async function sendRequest(request: WebRequest, options?: WebRequestOptions): Promise<WebResponse> {
+    let i = 0;
+    let retryCount = options && options.retryCount ? options.retryCount : 5;
+    let retryIntervalInSeconds = options && options.retryIntervalInSeconds ? options.retryIntervalInSeconds: 5;
+    let retriableErrorCodes = options && options.retriableErrorCodes ? options.retriableErrorCodes : ["ETIMEDOUT"];
```
Comment (bishal-pdMSFT) [inline Tasks/Common/azure-arm-rest/webClient.ts]: is there an error code for 502? that is something we can also think of for retry

File: Tasks/Common/azure-arm-rest/webClient.ts [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+
+    while (true) {
+        try {
+            return await sendReqeustInternal(request);
+        }
+        catch (error) {
+            if (retriableErrorCodes && retriableErrorCodes.indexOf(error.code) != -1 && ++i < retryCount) {
```
Comment (bishal-pdMSFT) [inline Tasks/Common/azure-arm-rest/webClient.ts]: first check is redundant, retriableErrorCodes is always initialized above

Comment (bishal-pdMSFT) [inline Tasks/Common/azure-arm-rest/webClient.ts]: we should check for error.code not being null

File: Tasks/Common/azure-arm-rest/webClient.ts [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+                tl.debug(util.format("Encountered a retriable error:%s. Message: %s.", error.code, error.message));
```
Comment (bishal-pdMSFT) [inline Tasks/Common/azure-arm-rest/webClient.ts]: debug or console log?

File: Tasks/Common/azure-arm-rest/webClient.ts [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+                await sleepFor(retryIntervalInSeconds);
+            }
+            else {
+                throw error;
```
Comment (bishal-pdMSFT) [inline Tasks/Common/azure-arm-rest/webClient.ts]: just `throw` does not work?

Comment (sachinma) [inline Tasks/Common/azure-arm-rest/webClient.ts]: AFAIK it is equivalent to throw ex of C#

File: Tasks/Common/azure-arm-rest/webClient.ts (hunk 7 above)
Comment (sachinma) [inline Tasks/Common/azure-arm-rest/webClient.ts]: i think it is fine to have tl.debug here.

File: Tasks/Common/azure-arm-rest/webClient.ts (hunk 6 above)
Comment (sachinma) [inline Tasks/Common/azure-arm-rest/webClient.ts]: If it is null, indexOf should return -1

File: Tasks/Common/azure-arm-rest/webClient.ts (hunk 4 above)
Comment (sachinma) [inline Tasks/Common/azure-arm-rest/webClient.ts]: Leaving it as such for now

File: Tasks/Common/azure-arm-rest/webClient.ts (hunk 5 above)
Comment (sachinma) [inline Tasks/Common/azure-arm-rest/webClient.ts]: Lets add it based on our observations later

File: Tasks/Common/azure-arm-rest/webClient.ts (hunk 6 above)
Comment (sachinma) [inline Tasks/Common/azure-arm-rest/webClient.ts]: removed

**Pull request merged**